FaunaDB.Client/Types/Api21/RecursiveRef.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using FaunaDB.Query;
+using Newtonsoft.Json;
+
+namespace FaunaDB.Types.Api21
+{
+    public abstract class ObjectRefV : RefV
+    {
+        public string Id { get { return Value; } }
+        public ObjectRefV Class { get; }
+        public ObjectRefV Database { get; }
+
+        protected ObjectRefV(string id, ObjectRefV cls, ObjectRefV db) : base(id)
+        {
+            Class = cls;
+            Database = db;
+        }
+
+        protected internal override void WriteJson(JsonWriter writer)
+        {
+            var props = UnescapedObject.With(
+                "id", Id,
+                "class", Class,
+                "database", Database
+            );
+
+            writer.WriteObject("@ref", props);
+        }
+
+        public override bool Equals(Expr v)
+        {
+            if (v == null || GetType() != v.GetType())
+                return false;
+
+            var w = v as ObjectRefV;
+
+            if (!Id.Equals(w.Id))
+                return false;
+
+            if (Class != null && !Class.Equals(w.Class))
+                return false;
+
+            if (Database != null && !Database.Equals(w.Database))
+                return false;
+
+            return true;
+        }
+
+        protected override int HashCode() =>
+            new { Id, Class, Database }.GetHashCode();
+
+        public override string ToString() =>
+            $"{GetType().Name}(id: \"{Id}\", class: {Option.Of(Class)}, database: {Option.Of(Database)})";
+    }
+
+    public class DatabaseV : ObjectRefV
+    {
+        public DatabaseV(string id, DatabaseV database = null) : base(id, ClassV.BuiltIn.DATABASES, database)
+        { }
+    }
+
+    public class ClassV : ObjectRefV
+    {
+        internal ClassV(string builtin) : base(builtin, null, null)
+        { }
+
+        public ClassV(string id, DatabaseV database = null) : base(id, BuiltIn.CLASSES, database)
+        { }
+
+        public static class BuiltIn
```

Comment:
The idea was to create some sort of dsl like `ClassV.BuiltIn.CLASSES`.

I don't know if it will be used by the users, theoretically the users could use it in `SelfV` and `InstanceV` but doesn't make sense to use with other types since we have specific types like `KeyV` and `FunctionV`